### PR TITLE
 Add $darken_ Utility Functions for Darkening Palette Colors 

### PR DESCRIPTION
**Pull request by shivam-daksh**
## Description
This PR introduces utility functions `$darken1`, `$darken2`, and `$darken3` to darken palette colors for use in button hover states and other components. These utilities are added to the `KThemePlugin` and are now available globally across all Vue components.

#### Issue addressed
#726 

### Before/after screenshots

image

### New Utility Functions:
- **`$darken1(color)`**: Slightly darkens the provided color.
- **`$darken2(color)`**: Moderately darkens the provided color.
- **`$darken3(color)`**: Heavily darkens the provided color.

These functions are defined in a new file, `darkenUtils.js`, which handles the color manipulation logic.
### Integration:
- **`KThemePlugin.js`**:
  - Imported the darken utility functions from `darkenUtils.js`.
  - Registered these utilities on the Vue prototype as `$darken1`, `$darken2`, and `$darken3`, making them globally accessible.

- **Example Usage in Vue Components**:
  ```javascript
  computed: {
    styles() {
      return {
        backgroundColor: this.$darken1(this.$themePalette.red.v_1100),
      };
    }
  }

## (optional) Implementation notes

### At a high level, how did you implement this?
<!-- Briefly describe how this works -->

### Does this introduce any tech-debt items?
<!-- List anything that will need to be addressed later -->

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## After review

- [ ] The changelog item has been pasted to the `CHANGELOG.md`

## Comments
<!-- Any additional notes you'd like to add -->

### CODE REVIEW

**Comment by shivam-daksh** [comment]
Hi @MisRob, Please review the PR.

**Comment by MisRob** [comment]
Thanks @shivam-daksh, the approach looks well to me overall! And very helpful PR description. I will loop in my colleagues for a full review - @ozer550 and/or @AllanOXDi, would you please have a look and also tried to test these utilities on a few colors, here from KDS or even Kolibri? The linked issue contains all the expectations. Thank you :)

File: lib/styles/darkenColors.js
Comment (MisRob) [inline lib/styles/darkenColors.js:1]: @shivam-daksh `'color'` package will need to be added to the `package.json`. Also then please run `yarn install` and commit both `package.json` and `yarn.lock`. Without it, it won't work for others :) Thanks!

Comment (shivam-daksh) [reply lib/styles/darkenColors.js:1]: Hey @MisRob, I first tried `npm install color` and `yarn add color` , both doesn't worked. 
So, I added `"color": "4.2.3"` in **package.json** file in `dependencies` manually and again fired `yarn install`. But it is giving me errors that my node version should be greater than `">=12.5.0"` however I'm using 10.24.1 as per documentation (10.x required).

image

Comment (MisRob) [reply lib/styles/darkenColors.js:1]: Your development environment is set up correctly. This means that this newest `color` package version requires Node >= 12.5.0. 

We can't upgrade our Node quite yet so we need to stay with Node v10. Could you try to install the highest version of the `color` package that still works with Node 10? It seems they restricted it in https://github.com/Qix-/color/releases/tag/4.2.1 so https://github.com/Qix-/color/releases/tag/4.2.0 may be a good candidate to try out. After you installed, please check that your implementation still works as expected.

Comment (MisRob) [reply lib/styles/darkenColors.js:1]: I'm not sure if this will work though because looking more at the changelogs, it seems that in `4.2.1` they only started communicating about the requirement for Node v12 that was there for longer time before.. https://github.com/Qix-/color/pull/236.

Comment (MisRob) [reply lib/styles/darkenColors.js:1]: Well, let's see if we can figure something out!

If not, we could wait for merging this PR until our Node v18 upgrade https://github.com/learningequality/kolibri-design-system/pull/645 is ready :)

Comment (shivam-daksh) [reply lib/styles/darkenColors.js:1]: Till then, I'm trying the other versions of `color` that are compatible with node@10.

Comment (shivam-daksh) [reply lib/styles/darkenColors.js:1]: Hey @MisRob , `color` version 3.2.1 is working fine as expected. Shoult I go with it?

Comment (MisRob) [reply lib/styles/darkenColors.js:1]: Great, thanks for researching that. Yes, let's give it a try. Could you pin it to this exact version in the package.json? So that we don't upgrade by mistake until we have Node 18.

File: package.json
Comment (shivam-daksh) [inline package.json:33]: @MisRob , I've pinned the exact version of color.

**Comment by AllanOXDi** [comment]
HI @shivam-daksh Thanks for your work here. I am taking look by testing it in few places

**Review by AllanOXDi — APPROVED** [review]
Thanks @shivam-daksh , LGTM!

**Review by shivam-daksh — COMMENTED** [review]
looks fine

**Comment by shivam-daksh** [comment]
Hey, @AllanOXDi  and @MisRob , There was a issue persisting `check if changelog.md is changed` , So I tried to update the changes in `CHANGELOG.md` manually. I thought it may disturb the workflows, so I reverted the changes. Waiting for your guidance and approval.

**Comment by AllanOXDi** [comment]
Hi @shivam-daksh  All you need is to add this to the changelog.  Something like

```
 [#728]
  - **Description:** [add the description of your PR  here]
  - **Products impact:**  None 
  - **Addresses:** https://github.com/learningequality/kolibri-design-system/issues/726
  - **Components:**  
  - **Breaking:** N0
  - **Impacts a11y:** No
  - **Guidance:**
  [#728]https://github.com/learningequality/kolibri-design-system/pull/728
```

**Comment by shivam-daksh** [comment]
Hey @AllanOXDi , I've changed the changelog file. Please review this again.

**Comment by AllanOXDi** [comment]
Thanks! look better. But you still need to add the description. Here is an example https://github.com/learningequality/kolibri-design-system/pull/722/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR8. I am sorry I was not clear.
If you could also resolve the conflicts-  would be good

**Comment by shivam-daksh** [comment]
@AllanOXDi , Do I've to change other files in documentation as given in  [this example](https://github.com/learningequality/kolibri-design-system/pull/722/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR8 ) or I've to change just `CHANGELOG.md`?

**Comment by AllanOXDi** [comment]
> @AllanOXDi , Do I've to change other files in documentation as given in [this example](https://github.com/learningequality/kolibri-design-system/pull/722/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR8) or I've to change just `CHANGELOG.md`?

No! What you did was good. just supposed to add the description of your work under the description

**Comment by AllanOXDi** [comment]
Looks good! lets now just resolve the merge conflicts

**Comment by shivam-daksh** [comment]
> Looks good! lets now just resolve the merge conflicts

@AllanOXDi, It's done.

**Comment by AllanOXDi** [comment]
Good to go!

**Comment by AllanOXDi** [comment]
> > Looks good! lets now just resolve the merge conflicts
> 
> @AllanOXDi, It's done.

Thanks!

**Comment by MisRob** [comment]
Thanks both @shivam-daksh and @AllanOXDi, and @shivam-daksh thanks for you first contribution. We appreciate it. I will merge soon.

FYI, I made use of one of my better screens and tweaked the values to correspond a bit more closely to resulting hex values mentioned in the issue.

Important part of acceptance criteria is documentation, however we typically do not require volunteers to write documentation, so I jotted it down and am pushing that as well:

Screenshot from 2024-08-21 15-33-03

**Review by MisRob — APPROVED** [review]

**Comment by MisRob** [comment]
Opened related issue https://github.com/learningequality/kolibri-design-system/issues/740

**Comment by MisRob** [comment]
@shivam-daksh if you're still interested in the follow-up issues

- https://github.com/learningequality/kolibri/issues/12553
- https://github.com/learningequality/studio/issues/4634

now it's the right time :)

Just message us in those issues.

**Comment by shivam-daksh** [comment]
Hi @MisRob, Thanks a lot for the feedback and tweaks! I appreciate you taking the time to document the darken utilities as well. It’s been a great learning experience. Although it was my second contribution (my first contribution was [Links leading outside Studio need to have a pop out icon](https://github.com/learningequality/studio/pull/4622). I would love to working on those issues.

**Comment by MisRob** [comment]
Lovely! You will need to post comment there, otherwise we can't assign.

Get some rest too ;)

**Comment by MisRob** [comment]
@shivam-daksh Lovely! You will need to post comment there, otherwise we can't assign. Get some rest too ;)

**Pull request merged**
